### PR TITLE
[RFC] Add experimental version using std::thread

### PR DESCRIPTION
**Pull request by mrry**
**TL;DR:** Would you be open to a version of `multi_thread_gemm.h` that uses `std::thread` instead of pthreads? We want to build TensorFlow on Windows, and it doesn't have pthreads, so we can't use gemmlowp at the moment.

I've been porting TensorFlow to Windows, and the missing gemmlowp support accounts for the remaining TF ops that don't build on Windows (i.e. the quantized flavors of ops). From a look at the codebase, the biggest thing holding us back is the dependency on pthreads. I made an experimental copy of `internal/multi_thread_gemm_no_pthread.h` that uses `std::thread` instead, and it seems to work for running `test/benchmark.cc` at least.

The performance seems similar:

### pthread baseline
```
$ gcc -O3 test/benchmark.cc -I. -std=c++11 -lstdc++ -lpthread
$ ./a.out 
Benchmarking small model GEMMs...
running for 10 seconds...
Graph latency (over 35 iterations):
  Best:             0.0344943s
  Worst:            0.0500524s
  Mean:             0.0381021s
  25% trimmed mean: 0.036896s
  Mean of 10% best: 0.0346334s
Benchmarking typical GoogLeNet GEMMs...
running for 20 seconds...
Graph latency (over 69 iterations):
  Best:             0.183233s
  Worst:            0.375618s
  Mean:             0.24s
  25% trimmed mean: 0.233365s
  Mean of 10% best: 0.188828s
Benchmarking default mode (typically multi-threaded)...
10x10x10 : 0.7204 GFlops/s                      
20x20x20 : 1.502 GFlops/s
30x30x30 : 2.813 GFlops/s
40x40x40 : 3.107 GFlops/s
50x50x50 : 2.543 GFlops/s
60x60x60 : 8.154 GFlops/s
64x256x147 : 15.17 GFlops/s
100x100x1 : 0.426 GFlops/s
100x100x100 : 10.52 GFlops/s
100x1000x100 : 12.71 GFlops/s
1000x1000x1 : 1.272 GFlops/s
1000x1000x10 : 13.71 GFlops/s
1000x1000x100 : 23.42 GFlops/s
1000x1000x1000 : 22.14 GFlops/s

Benchmarking single-threaded mode...
10x10x10 : 0.6435 GFlops/s                      
20x20x20 : 1.462 GFlops/s
30x30x30 : 2.456 GFlops/s
40x40x40 : 2.784 GFlops/s
50x50x50 : 2.406 GFlops/s
60x60x60 : 3.167 GFlops/s
64x256x147 : 3.876 GFlops/s
100x100x1 : 0.3945 GFlops/s
100x100x100 : 3.433 GFlops/s
100x1000x100 : 4.164 GFlops/s
1000x1000x1 : 0.4561 GFlops/s
1000x1000x10 : 2.634 GFlops/s
1000x1000x100 : 4.398 GFlops/s
1000x1000x1000 : 4.488 GFlops/s
```

### std::thread version
```
$ gcc -O3 test/benchmark.cc -I. -std=c++11 -DGEMMLOWP_USE_STD_THREAD -lstdc++ -lpthread
$ ./a.out 
Benchmarking small model GEMMs...
running for 10 seconds...
Graph latency (over 36 iterations):
  Best:             0.0344355s
  Worst:            0.0537341s
  Mean:             0.0384871s
  25% trimmed mean: 0.0370866s
  Mean of 10% best: 0.0345953s
Benchmarking typical GoogLeNet GEMMs...
running for 20 seconds...
Graph latency (over 72 iterations):
  Best:             0.184519s
  Worst:            0.30761s
  Mean:             0.230187s
  25% trimmed mean: 0.225073s
  Mean of 10% best: 0.188191s
Benchmarking default mode (typically multi-threaded)...
10x10x10 : 0.7238 GFlops/s                      
20x20x20 : 1.465 GFlops/s
30x30x30 : 2.772 GFlops/s
40x40x40 : 3.125 GFlops/s
50x50x50 : 2.56 GFlops/s
60x60x60 : 8.223 GFlops/s
64x256x147 : 13.77 GFlops/s
100x100x1 : 0.4266 GFlops/s
100x100x100 : 10.13 GFlops/s
100x1000x100 : 12.77 GFlops/s
1000x1000x1 : 2.325 GFlops/s
1000x1000x10 : 14.33 GFlops/s
1000x1000x100 : 15.77 GFlops/s
1000x1000x1000 : 25.45 GFlops/s

Benchmarking single-threaded mode...
10x10x10 : 0.6748 GFlops/s                      
20x20x20 : 1.348 GFlops/s
30x30x30 : 2.603 GFlops/s
40x40x40 : 2.902 GFlops/s
50x50x50 : 2.408 GFlops/s
60x60x60 : 2.94 GFlops/s
64x256x147 : 4.232 GFlops/s
100x100x1 : 0.3948 GFlops/s
100x100x100 : 3.435 GFlops/s
100x1000x100 : 3.984 GFlops/s
1000x1000x1 : 0.4831 GFlops/s
1000x1000x10 : 2.637 GFlops/s
1000x1000x100 : 4.398 GFlops/s
1000x1000x1000 : 4.561 GFlops/s
```

If this seems acceptable, what would be the most acceptable way to add this support to gemmlowp? (Presumably duplicating the code like I've done in the PR is *not* acceptable!)

### CODE REVIEW

**Comment by bjacob** [comment]
Good question.

First, I guess that the number of people who will want to run TensorFlow on Windows, with a quantized model, caring about performance and OK to run on multiple cores, is probably small. So for the purpose of porting TensorFlow to Windows, an easy way out is probably to just disable multi-threading in gemmlowp on Windows. I'd be perfectly happy with that.

If you do want multi-threaded gemmlowp on Windows, then indeed we need to figure out how to make this diff less intrusive / duplicating.

One way that would be very nice as far as I'm concerned, is if you could implement a minimal pthreads shim, exposing just the pthreads functionality that this file needs --- so we could pretend that we're always on a POSIX platform. Similar to how Chrome on Windows works by wrapping Direct3D in a OpenGL compatibility layer (ANGLE).

If implementing that small subset of pthreads on top of std::threads is not convenient for whatever reason, you'd have the option to implement it instead on top of traditional Windows threading functions (which I'm not familiar with). Windows is probably the only non-POSIX platform that matters, anyway.

Does that sound reasonable?

**Comment by bjacob** [comment]
Also --- if you're running benchmarks on x86, please build with -msse4. Otherwise you're getting the scalar path.

**Comment by bhack** [comment]
What is the problem of using c++11 threads?

**Comment by bjacob** [comment]
Primary concern: portability to mobile/embedded platforms. In principle c++11 threads is more standard/portable than pthreads; in practice, pthreads support is ubiquitous outside of Windows, while c++11 threads support varies a lot. Some C++ standard library implementations support it but with bad/unpredictable performance.

Android itself is navigating 3 different C++ standard library implementations (stlport is still used in legacy Android toolchains, libstdc++ is the present and libc++ the future), and there have been perf issues with c++11 threads on Android in the recent past, so I would like at least the transition to libc++ to be behind us before we get exposed to such Android toolchain details.

There is a secondary concern, which is that on mobile devices it may take some hacks to coerce the OS to keep CPU cores spinning and give time slices to gemmlowp worker threads (mobile OSes are special there because of how aggressive they are in spinning CPU cores down to conserve battery charge). See
https://github.com/google/gemmlowp/blob/master/internal/multi_thread_gemm.h#L100
I know that we could do the same with c++11 threads, but on a higher level, knowing that we need this kind of low-level control tips the scales in favor of pthreads, which is the lower-level C API that maps more directly to system calls, compared to the more abstract C++11 threads.

**Comment by bhack** [comment]
Probably caring library implementation of [yeild](http://en.cppreference.com/w/cpp/thread/yield) for android could be enough. In the mean time you could consider also adding [pthread4w] (https://sourceforge.net/p/pthreads4w/wiki/Home/) dependecy on Win

**Comment by bjacob** [comment]
To be clear, I think that what we need here is less than 1% of a full pthreads implementation, probably a 50-line shim.

Regarding the low-level control that we need, it's more like the opposite of yield --- sometimes we want to keep our timeslice in case we might have to start a new worker task very soon.

**Comment by bhack** [comment]
So how do you are balancing this with power consumption? I think that we are generally  talking of devices powered with batteries.

**Comment by bjacob** [comment]
The idea is that this allows the GEMM to complete 2x quicker, so it more than makes up for the small amount of extra time (a few ms) that cores have been kept spinning.
But it certainly is a tough balancing act.
What I learned more recently is that on many mobile devices, thermal issues mean what we often cannot afford to run any multi-threaded workload --- regardless of battery charge considerations. So in practice I have to say that gemmlowp multithreading is not that much used, in real mobile applications.

**Comment by bjacob** [comment]
Oh, and the other thing that changed is that NNs have shrunk a lot. So the GEMMs that we do to perform NN inference have become very small. Say 150x150 is a typical matrix size. Thus, a big chunk of the GEMMs that we have to perform simply do not benefit from multi-threading at all. For example, a 150x150x150 GEMM involves about 3M scalar multiply-add operations. If we operate at 3G scalar multiply-adds per second (a typical performance level for single-threaded GEMM on that kind of matrix size on a typical ARM CPU) then that means that we complete this GEMM in about 1 ms. Clearly that's too short a time to benefit from anything but the most carefully tailored multi-threading primitives.

**Comment by bhack** [comment]
Are Mali and Adreno are out of scope? Generally this kind of devs has host unfied memory.

**Comment by bjacob** [comment]
All mobile SoC's that I know about have unified memory, yes.
(Not sure how that, and generally GPUs (Adreno mentioned here), is relevant to this PR)

**Comment by bhack** [comment]
GPU it is not related to this PR but was related to the performance power consumption digression.

**Comment by mrry** [comment]
Sorry for the extended delay in replying to this! I'm going to close this PR for now, since this version of the code isn't in a mergeable state. I tried to implement a pthread-like shim for Windows, but it got ugly pretty fast. 

It'd be great to figure out the best way to make gemmlowp build on Windows, as the quantized ops are the main missing piece in TensorFlow's cross-platform compatibility story. However, as nobody has specifically demanded them (yet!), it's not worth making invasive changes to gemmlowp to get there.

/cc @petewarden since we were just chatting about this.

**Comment by bhack** [comment]
We are also using gemmlowp as upstream in tinydnn  /cc @wangyisa

**Comment by bjacob** [comment]
Again: there is a very simple avenue to running on windows, it is to run single-threaded without pthreads. This has been how gemmlowp has been used anyway even on platforms that have pthreads, i.e. gemmlowp multi-threading hasn't been as useful as initially hoped. (so far, anyway).
